storage/secondary.go

```diff
@@ -0,0 +1,107 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"sync"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
+)
+
+// secondaryQuerier is a wrapper that allows a querier to be treated in a best effort manner.
+// This means that an error on any method returned by Querier except Close will be returned as a warning,
+// and the result will be empty.
+//
+// Additionally, Querier ensures that if ANY SeriesSet returned by this querier's Select failed on an initial Next,
+// All other SeriesSet will be return no response as well. This ensures consistent partial response strategy, where you
+// have either full results or none from each secondary Querier.
+// NOTE: This works well only for implementations that only fail during first Next() (e.g fetch from network). If implementation fails
+// during further iterations, set will panic. If Select is invoked after first Next of any returned SeriesSet, querier will panic.
+//
+// Not go-routine safe.
+// NOTE: Prometheus treats all remote storages as secondary / best effort.
+type secondaryQuerier struct {
+	genericQuerier
+
+	once      sync.Once
+	done      bool
+	asyncSets []genericSeriesSet
+}
+
+func newSecondaryQuerierFrom(q Querier) genericQuerier {
+	return &secondaryQuerier{genericQuerier: newGenericQuerierFrom(q)}
+}
+
+func newSecondaryQuerierFromChunk(cq ChunkQuerier) genericQuerier {
+	return &secondaryQuerier{genericQuerier: newGenericQuerierFromChunk(cq)}
+}
+
+func (s *secondaryQuerier) LabelValues(name string) ([]string, Warnings, error) {
+	vals, w, err := s.genericQuerier.LabelValues(name)
+	if err != nil {
+		return nil, append([]error{err}, w...), nil
+	}
+	return vals, w, nil
+}
+
+func (s *secondaryQuerier) LabelNames() ([]string, Warnings, error) {
+	names, w, err := s.genericQuerier.LabelNames()
+	if err != nil {
+		return nil, append([]error{err}, w...), nil
+	}
+	return names, w, nil
+}
+
+func (s *secondaryQuerier) createFn(asyncSet genericSeriesSet) func() (genericSeriesSet, bool) {
+	s.asyncSets = append(s.asyncSets, asyncSet)
+	curr := len(s.asyncSets) - 1
+	return func() (genericSeriesSet, bool) {
+		s.once.Do(func() {
+			// At first create invocation we iterate over all sets and ensure its Next() returns some value without
+			// errors. This is to ensure we support consistent partial failures.
+			var errs tsdb_errors.MultiError
+			for i, set := range s.asyncSets {
+				if set.Next() {
+					continue
+				}
+				// Failed or exhausted set.
+				ws := set.Warnings()
+				if err := set.Err(); err != nil {
+					errs.Add(err)
+					ws = append([]error{err}, ws...)
+				}
+				s.asyncSets[i] = warningsOnlySeriesSet(ws)
+			}
+
+			if errs.Err() != nil {
+				// One failed, ensure all sets returns nothing. (All or nothing logic).
+				for i, set := range s.asyncSets {
+					s.asyncSets[i] = warningsOnlySeriesSet(set.Warnings())
```

Comment:
Hmm, would this mean the warnings will be duplicated many times to the user? Might we want to do a noop series set for the others instead?

Reply:
This is very valid point, let's do that and ensure on our new tests. I remember I wanted to do it , but forgot.